src/gluonts/model/forecast_generator.py

```diff
@@ -241,3 +262,114 @@ def __call__(
                     info=batch["info"][i] if "info" in batch else None,
                 )
             assert i + 1 == len(batch[FieldName.FORECAST_START])
+
+
+class ForecastBatch:
+    @property
+    def batch_size(self) -> int:
+        raise NotImplementedError
+
+    @property
+    def mean(self) -> np.ndarray:
+        raise NotImplementedError
+
+    def quantile(self, q: Union[float, str]) -> np.ndarray:
+        raise NotImplementedError
+
+
+@dataclass
+class DistributionForecastBatch(ForecastBatch):
+    start_date: list
+    item_id: Optional[list]
+    info: Optional[list]
+    distr_output: Any  # TODO fix
+    distr_args: list
+
+    def __post_init__(self):
+        self.distr = self.distr_output.distribution(*self.distr_args)
+
+    def __iter__(self) -> Iterator[Forecast]:  # TODO fix
+        distributions = [
+            self.distr_output.distribution(*u)
+            for u in recursively_zip_arrays(self.distr_args)
+        ]
+
+        for i, distr in enumerate(distributions):
+            yield make_distribution_forecast(
+                distr,
+                start_date=self.start_date[i],
+                item_id=self.item_id[i] if self.item_id is not None else None,
+                info=self.info[i] if self.info is not None else None,
+            )
+
+    @property
+    def mean(self) -> np.ndarray:
+        return to_numpy(self.distr.mean())
+
+    def quantile(self, q: Union[float, str]) -> np.ndarray:
+        return to_numpy(self.distr.quantile(q))
+
+
+@dataclass
+class SampleForecastBatch(ForecastBatch):
+    start_date: list
+    item_id: Optional[list]
+    info: Optional[list]
+    sample_batch: np.ndarray
+
+    def __post_init__(self):
+        self._sorted_samples_value = None
```

Comment:
Why not just directly assign sorted samples here? Do we require it to be lazy?

Reply:
It may be premature optimization, but as long as you don't need to access quantiles, there's no need for sorting